Fire teleportation events when teleporting an entity

syldium opened the pull request:
# Description
The `Entity#teleport` method now fires the `EntityTeleportEvent` if it's not a player. If it's a player who goes to another dimension, this method will also fire the `PlayerChangedWorldEvent` afterward.

Some preconditions have been added, namely for non-nullable arguments. Some checks prevent weird states, like if the entity is removed or dead.

The `Entity#teleport` method now also makes a defensive copy of the location to prevent external mutations. The entity is the only owner of this location.

# Checklist
The following items should be checked before the pull request can be merged.
- [x] Unit tests added.
- [x] Code follows existing (local) code format.

# For maintainer
When a PR is approved, the maintainer should label this PR with `release/*`.

- If the PR fixes a bug in MockBukkit, update the patch version. (if the current version is `0.4.1`, the new version should be `0.4.2`)
- If the PR adds a new feature to MockBukkit, update minor version. (if the current version is `0.4.1`, the new version should be `0.5.0`)

Note that a PR that fixes an `UnimplementedOperationException` should be considered a new version and not a bugfix.

syldium committed:
Fire PlayerChangedWorldEvent when teleporting a player to a different world

syldium committed:
Fire EntityTeleportEvent when teleporting an entity

Add checks if the location has a world and if the entity is removed or dead.

Insprill commented on src/test/java/be/seeseemelk/mockbukkit/entity/PlayerMockTest.java:
```diff
@@ -1363,6 +1368,27 @@ void testPlayerTeleport_WithoutCause_EventFired()
+	{
+		TestPlugin plugin = MockBukkit.load(TestPlugin.class);
+		World from = player.getWorld();
+		World[] playerWorldWhenEvent = new World[1];
```
Is there a particular reason this is an array?

syldium replied:
The variable needs to be effectively final to be accessed from within the inner class. The array is, or it could be any holder object like an `AtomicReference` but that would be overkill.

syldium replied:
In fact, the assertion can be made in the inner class. This removes the array.

Insprill replied:
Exceptions are just ignored in events so the new way can't fail. However, this is an issue with MockBukkit, so I'm going to mark this resolved and fix it in its own PR.

syldium committed:
Accept teleportation to null worlds for entities only

Starting from CraftBukkit#991aeda1218e72e9237942071e7ea55cc77b541c,
entities can now be teleported to a location without a defined world.

syldium committed:
Move the assertion into the inner class

thelooter committed:
Merge branch 'v1.19' into feature/teleport-events

Insprill commented on src/main/java/be/seeseemelk/mockbukkit/entity/EntityMock.java:
```diff
@@ -259,18 +266,35 @@ public void removeMetadata(@NotNull String metadataKey, @NotNull Plugin owningPl
+		{
+			return false;
+		}
+
```
```suggestion
		//todo: Add passenger logic

Insprill commented on src/main/java/be/seeseemelk/mockbukkit/entity/PlayerMock.java:
```diff
@@ -2742,7 +2743,13 @@ public int getPing()
+		{
+			return false;
+		}
 
```
```suggestion
		//todo: Add passenger logic

Insprill commented on src/main/java/be/seeseemelk/mockbukkit/entity/PlayerMock.java:
```diff
@@ -2742,7 +2743,13 @@ public int getPing()
+		}
 
 		PlayerTeleportEvent playerTeleportEvent = new PlayerTeleportEvent(this, getLocation(), location, cause);
 		Bukkit.getPluginManager().callEvent(playerTeleportEvent);
```
This and the `#isCancelled` check can be replaced with `Event#callEvent`

Insprill commented on src/main/java/be/seeseemelk/mockbukkit/entity/PlayerMock.java:
```diff
@@ -2752,7 +2759,13 @@ public boolean teleport(@NotNull Location location, @NotNull PlayerTeleportEvent
 			return false;
 		}
 
```
Any open inventories should be closed ([source](https://hub.spigotmc.org/stash/projects/SPIGOT/repos/craftbukkit/browse/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java#827-830))

Insprill committed:
Add notes to implement passenger logic in the future

Insprill committed:
Merge branch 'v1.19' into feature/teleport-events

thelooter committed:
Merge branch 'v1.19' into feature/teleport-events

syldium committed:
Refactor with Event#callEvent

Co-authored-by: Pierce Thompson <52843537+Insprill@users.noreply.github.com>

syldium committed:
Close open inventories when teleporting a player

syldium committed:
Add comments

And re-trigger a build

thelooter committed:
Merge branch 'v1.19' into feature/teleport-events

Insprill committed:
Merge branch 'v1.19' into feature/teleport-events

syldium committed:
Fix failing test

thelooter committed:
Merge branch 'v1.19' into feature/teleport-events

thelooter commented:
@Insprill Is this good to go

thelooter committed:
Merge branch 'v1.19' into feature/teleport-events

Insprill committed:
Merge branch 'v1.19' into feature/teleport-events

Insprill committed:
Merge branch 'v1.19' into feature/teleport-events

Insprill reviewed (approved):
LGTM
